File: fixtures/csharp/controller-service-provider/src/Payments.Core/PaymentService.cs
using Chakra.Payments.Provider;
using Chakra.Payments.Shared;

namespace Chakra.Payments.Service;

public partial class PaymentService<TProvider>
    where TProvider : IPaymentProvider
{
    private readonly TProvider provider;

    public PaymentService(TProvider provider)
    {
        this.provider = provider;
    }

    public async Task<string> RefundAsync(int amountCents)
    {
        Shared.RecordEvent("refund");
        Shared.SharedUniqueTarget("payment".NormalizePayment());
        return await provider.RefundAsync(amountCents);
    }

    public static PaymentService<StripeProvider> BuildPaymentService(string apiKey)
    {
        return new PaymentService<StripeProvider>(new StripeProvider(apiKey));
    }
}

File: fixtures/java/controller-service-provider/src/main/java/chakra/payments/service/PaymentService.java
package chakra.payments.service;

import chakra.payments.provider.PaymentProvider;
import chakra.payments.provider.StripeProvider;
import chakra.payments.shared.Shared;

import static chakra.payments.shared.Shared.recordEvent;

/** Service coordinating refunds through the configured provider. */
public class PaymentService {

    private final PaymentProvider provider;

    public PaymentService(PaymentProvider provider) {
        this.provider = provider;
    }

    public boolean refund(long amountCents) {
        Shared.sharedUniqueTarget();
        return this.provider.refund(amountCents);
    }

    public static PaymentService buildPaymentService(String apiKey) {
        recordEvent("build");
        return new PaymentService(new StripeProvider(apiKey));
    }
}
